### PR TITLE
Add email notification toggle

### PR DESCRIPTION
**Pull request by ouchadam**
#2243 Email notification toggles

- Adds a `Email notifications` category to `settings/notifications` which allows for individually toggling email notifications for all associated emails
    - When the account has no associated emails, the category contains a link to the `Email and Phone numbers` screen 
    - `addEmailPusher` functionality is added to the `matrix-sdk PushersService` which allows an email address to be registered or unregistered for notifications  
- Adds a `REMOVE` button to the `settings/advanced settings/notification targets`
- Renames a few `AddHttpPusher` classes to `AddPusher` as they were able to be repurposed for emails 

* personal data has been manually overridden and replaced with [redacted]

| BEFORE | NO EMAILS AVAILABLE | EMAIL PRESENT | LIGHT - MULTIPLE EMAILS |
| --- | --- | --- | --- |
|account-before|account-no-emails-registered|account-emails-notification-disabled|account-light-multiple

| NAVIGATING EMPTY | REMOVING | ADDING | ERROR HANDLING |
| --- | --- | --- | --- |
|account-no-emails-clickthrough|account-removing-email-pusher|account-registering-email|account-error-handling

### CODE REVIEW

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt [hunk 1]
```diff
@@ -66,27 +66,39 @@ internal class DefaultPushersService @Inject constructor(
...
-                appId = appId,
-                appDisplayName = appDisplayName,
-                deviceDisplayName = deviceDisplayName,
-                profileTag = profileTag,
-                lang = lang,
-                data = JsonPusherData(url, EVENT_ID_ONLY.takeIf { withEventIdOnly }),
-                append = append)
+                               withEventIdOnly: Boolean
+    ) = addPusher(
+            JsonPusher(
```
Comment (ouchadam) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt:71]: there's a lot of duplication here caused by the amount of parameters, the next PR coming up will address this by introducing a dedicated `HttpPusher` data class

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushers/PushersService.kt [hunk 2]
```diff
@@ -80,9 +97,10 @@ interface PushersService {
                          eventId: String)
 
     /**
-     * Remove the http pusher
+     * Remove the pusher
+     * When the appId is m.email then an email address is expected in the pushkey
      */
-    suspend fun removeHttpPusher(pushkey: String, appId: String)
+    suspend fun removePusher(pushkey: String, appId: String)
```
Comment (ouchadam) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushers/PushersService.kt:103]: renamed because _technically_ this function can remove email and http pushers

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushers/PushersService.kt [hunk 3]
```diff
@@ -64,6 +62,25 @@ interface PushersService {
...
+    /**
+     * Add a new Email pusher.
+     * Ref: https://matrix.org/docs/spec/client_server/latest#post-matrix-client-r0-pushers-set
+     *
+     * @param email             The email address to send notifications to.
+     * @param lang              The preferred language for receiving notifications (e.g. "en" or "en-US").
+     * @param emailBranding     The branding placeholder to include in the email communications.
+     * @param appDisplayName    A human readable string that will allow the user to identify what application owns this pusher.
+     * @param deviceDisplayName A human readable string that will allow the user to identify what device owns this pusher.
+     * @return                  A work request uuid. Can be used to listen to the status
```
Comment (ouchadam) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushers/PushersService.kt:74]: the next PR will remove the reliance on the `WorkManager` by switching to a  `suspend fun`, allowing us to handle and simplify the registration errors

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt [hunk 4]
```diff
@@ -66,27 +66,39 @@ internal class DefaultPushersService @Inject constructor(
...
+                    append = append
+            )
+    )
 
-        val params = AddHttpPusherWorker.Params(sessionId, pusher)
+    override fun addEmailPusher(email: String, lang: String, emailBranding: String, appDisplayName: String, deviceDisplayName: String) = addPusher(
+            JsonPusher(
+                    pushKey = email,
+                    kind = "email",
+                    appId = "m.email",
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt:88]: Maybe create public const val for `kind` "email" and "http" (so in `PusherService`) and private const `appId` "m.email" ?

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt [hunk 5]
```diff
@@ -95,7 +107,14 @@ internal class DefaultPushersService @Inject constructor(
         return request.id
     }
 
-    override suspend fun removeHttpPusher(pushkey: String, appId: String) {
+    private fun JsonPusher.validateParameters() {
+        // Do some parameter checks. It's ok to throw Exception, to inform developer of the problem
+        if (this.pushKey.length > 512) throw InvalidParameterException("pushkey should not exceed 512 chars")
+        if (this.appId.length > 64) throw InvalidParameterException("appId should not exceed 64 chars")
+        this.data?.url?.let { url -> if ("/_matrix/push/v1/notify" !in url) throw InvalidParameterException("url should contain '/_matrix/push/v1/notify'") }
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt:114]: Generally, if it is not necessary, I do not like to explicitly use `this.`. We can discuss it if you want to :)

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt [hunk 6]
```diff
@@ -66,27 +66,39 @@ internal class DefaultPushersService @Inject constructor(
...
+                    lang = lang,
+                    appDisplayName = appDisplayName,
+                    deviceDisplayName = deviceDisplayName,
+                    data = JsonPusherData(url, EVENT_ID_ONLY.takeIf { withEventIdOnly }),
+                    append = append
+            )
+    )
 
-        val params = AddHttpPusherWorker.Params(sessionId, pusher)
+    override fun addEmailPusher(email: String, lang: String, emailBranding: String, appDisplayName: String, deviceDisplayName: String) = addPusher(
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt:84]: One param per line is better when there are a lot of parameters like here

Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt:84]: Also why there is no param `append` here?

File: vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt [hunk 7]
```diff
@@ -116,11 +120,48 @@ class VectorSettingsNotificationPreferenceFragment @Inject constructor(
...
 
+    private fun bindEmailNotifications() {
+        val emails = session.getEmailsWithPushInformation()
+        findPreference<VectorPreferenceCategory>(VectorPreferences.SETTINGS_EMAIL_NOTIFICATION_CATEGORY_PREFERENCE_KEY)?.let { category ->
+            if (emails.isEmpty()) {
+                val vectorPreference = VectorPreference(requireContext())
+                vectorPreference.title = resources.getString(R.string.settings_notification_emails_no_emails)
+                category.addPreference(vectorPreference)
+                vectorPreference.setOnPreferenceClickListener {
+                    interactionListener?.navigateToEmailAndPhoneNumbers()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt:137]: If I go this way and add an email to the account, when coming back here, the screen is not refreshed. Maybe force a refresh in onResume()

File: vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt [hunk 8]
```diff
@@ -116,11 +120,48 @@ class VectorSettingsNotificationPreferenceFragment @Inject constructor(
...
+                            pushManager.unregisterEmailPusher(emailPid.email)
+                        }
+                    }
+                    category.addPreference(pref)
+                }
+            }
+        }
+    }
+
+    private fun Session.getEmailsWithPushInformation(): List<Pair<ThreePid.Email, Pusher?>> {
```
[the remark this answers is not in the document]
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt:158]: Thanks for the explanation, but it will be more useful for a future reference if it is located in the code (doc of the extension).

File: vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt [hunk 9]
```diff
@@ -116,11 +120,48 @@ class VectorSettingsNotificationPreferenceFragment @Inject constructor(
...
+        val emailPushers = this.getPushers().filter { it.kind == "email" }
+        return this.getThreePids()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt:160]: Same remark about `this.`

File: vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt [hunk 10]
```diff
@@ -116,11 +120,48 @@ class VectorSettingsNotificationPreferenceFragment @Inject constructor(
...
+                        }
+                    }
+                    category.addPreference(pref)
+                }
+            }
+        }
+    }
+
+    private fun Session.getEmailsWithPushInformation(): List<Pair<ThreePid.Email, Pusher?>> {
+        val emailPushers = this.getPushers().filter { it.kind == "email" }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt:159]: Use the const you've just defined for "email" :)

File: vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt [hunk 11]
```diff
@@ -343,3 +384,19 @@ class VectorSettingsNotificationPreferenceFragment @Inject constructor(
                 }
     }
 }
+
+private fun SwitchPreference.setTransactionalSwitchChangeListener(scope: CoroutineScope, transaction: suspend (Boolean) -> Unit) {
+    this.setOnPreferenceChangeListener { switchPreference, isChecked ->
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt:389]: Useless `this.`

File: vector/src/main/java/im/vector/app/features/settings/push/PushGatewaysViewModel.kt [hunk 12]
```diff
@@ -70,10 +72,17 @@ class PushGatewaysViewModel @AssistedInject constructor(@Assisted initialState:
...
         when (action) {
-            is PushGatewayAction.Refresh -> handleRefresh()
+            is PushGatewayAction.Refresh      -> handleRefresh()
+            is PushGatewayAction.RemovePusher -> removePusher(action.pusher)
         }.exhaustive
     }
 
+    private fun removePusher(pusher: Pusher) {
+        viewModelScope.launch {
+            session.removePusher(pusher.pushKey, pusher.appId)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/settings/push/PushGatewaysViewModel.kt:82]: You have to protect this call with a try catch. If triggered in Airplane mode, it will crash the app. In case of Failure you can use ViewEvent to sent the failure and display it in a dialog (many examples exist in the code)

File: vector/src/main/res/layout/item_pushgateway.xml [hunk 13]
```diff
@@ -130,5 +129,11 @@
         android:textStyle=""
         tools:text="event_id_only" />
 
+    <Button
```
Comment (bmarty) [inline vector/src/main/res/layout/item_pushgateway.xml:132]: Format and URL are not relevant for Email, can you fix that please? (Not really part of the PR, but a good opportunity to make it right):
image

File: vector/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -1112,6 +1112,10 @@
     <string name="settings_notification_advanced">Advanced Notification Settings</string>
     <string name="settings_notification_by_event">Notification importance by event</string>
 
+    <string name="settings_notification_emails_category">Email notification</string>
+    <string name="settings_notification_emails_no_emails">To receive email with notification, please associate an email to your Matrix account</string>
+    <string name="settings_notification_emails_enable_for_email">Enable email notifications for %s</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml:1117]: Can you add a XML comment for translators, just above this line to tell them that the placeholder will be replaced by an email? There are other comment like that in this file. Thanks.

File: vector/src/main/res/values/strings.xml [hunk 15]
```diff
@@ -2245,6 +2249,7 @@
     <string name="push_gateway_item_device_name">session_name:</string>
     <string name="push_gateway_item_url">Url:</string>
     <string name="push_gateway_item_format">Format:</string>
+    <string name="push_gateway_item_remove">Remove</string>
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml:2252]: Maybe use the existing string R.string.remove (which should be renamed in the future with `action_remove`, for name coherency with other common actions), to avoid too many string duplication.

**Review by bmarty — CHANGES_REQUESTED** [review]
Good work thanks.
A few comments and remarks

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt (hunk 4 above)
Comment (ouchadam) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt:88]: sounds good, will do :+1:

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt (hunk 5 above)
Comment (ouchadam) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt:114]: happy to remove :+1:

File: vector/src/main/java/im/vector/app/features/settings/push/PushGatewaysViewModel.kt (hunk 12 above)
Comment (ouchadam) [reply vector/src/main/java/im/vector/app/features/settings/push/PushGatewaysViewModel.kt:82]: good point, will fix :+1: I had gotten used to debug pages taking shortcuts :sweat_smile:

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt (hunk 6 above)
Comment (ouchadam) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt:84]: `append` was removed as an option because it sounds like it has the side effect of causing other accounts which use the same email to lose email notifications https://github.com/matrix-org/matrix-react-sdk/pull/2727/files#diff-ec232520bf51337e5e6939b885d21f428ad6da3306c8e17a3ff660b2b341179dR165

File: vector/src/main/res/layout/item_pushgateway.xml (hunk 13 above)
Comment (ouchadam) [reply vector/src/main/res/layout/item_pushgateway.xml:132]: 1865e3c450483be7e43d0a399ba767493bbd4850

| HTTP | EMAIL |
| --- | --- |
|device-2021-09-23-125643|device-2021-09-23-125652

File: vector/src/main/java/im/vector/app/features/settings/push/PushGatewaysViewModel.kt (hunk 12 above)
Comment (ouchadam) [reply vector/src/main/java/im/vector/app/features/settings/push/PushGatewaysViewModel.kt]: 4c4f2fce74d3b0f0dc4fe6136b40b9e440e90116

offline-remove-pusher

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushers/PushersService.kt (hunk 2 above)
Comment (ouchadam) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushers/PushersService.kt:103]: I originally had that but it felt odd that `PushGateway` could still use the `removeHttpPusher` for both http and email pushers

the `PushersManager` has symmetrical register/unregister for http and email but that's a layer above the SDK

we could force the different flows based on appId

```kotlin
fun removePusher(pusher: Pusher) {
  when (pusher.appId) {
      Pusher.APP_ID_EMAIL -> session.removeEmailPusher(pusher.pushKey)
      else -> session.removeHttpPusher(pusher.pushKey, pusher.appId)
  }
}
``` 

what do you think?

File: vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt (hunk 7 above)
Comment (ouchadam) [reply vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt:137]: another option is to make use of `getThreePidsLive` which also has the benefit of updating the emails whilst in the screen, so far it's working well :crossed_fingers:

File: vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (ouchadam) [reply vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt]: doc added and updated to use a boolean 3a1cb1c07a02a3bd3a52f5f3bd0b0019bbb304b9

File: vector/src/main/res/values/strings.xml (hunk 15 above)
Comment (ouchadam) [reply vector/src/main/res/values/strings.xml]: 43eff8b85bbfa4c20f21204c9d3d05444b3b1770

File: vector/src/main/res/values/strings.xml (hunk 14 above)
Comment (ouchadam) [reply vector/src/main/res/values/strings.xml:1119]: 1d4c909b50b5b9f4ed8656ee71b340c3f7310cc3

File: vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt]: Perfect! Thanks

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushers/PushersService.kt (hunk 2 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushers/PushersService.kt:103]: The `PushersManager` is on the app side, so I would say its Api is less "important" than the SDK API `PushersService`. I agree that `removeHttpPusher` is not KO to remove email pushers.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt (hunk 6 above)
Comment (bmarty) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt]: It's sometimes difficult to decide between what we want the SDK to expose (as parameters for instance) and what we want the SDK to handle, so keep hidden from the client app. On one side the SDK should let the app do whatever it wants to, and on the other side we want the SDK to provide a high level service... So maybe add the parameter, with default value to `true` and with the same comment from https://github.com/matrix-org/matrix-react-sdk/pull/2727/files#diff-ec232520bf51337e5e6939b885d21f428ad6da3306c8e17a3ff660b2b341179dR165 in the doc of the fun `addEmailPusher`.
Thanks!

File: vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt (hunk 7 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/settings/notifications/VectorSettingsNotificationPreferenceFragment.kt:150]: far better, thanks!

File: vector/src/main/java/im/vector/app/features/settings/push/PushGatewaysViewModel.kt (hunk 12 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/settings/push/PushGatewaysViewModel.kt]: Thanks, crash in debug pages are also bad for the statistics :)

File: vector/src/main/res/layout/item_pushgateway.xml (hunk 13 above)
Comment (bmarty) [reply vector/src/main/res/layout/item_pushgateway.xml:132]: thanks!

File: vector/src/main/java/im/vector/app/core/extensions/ViewExtensions.kt [hunk 16]
```diff
@@ -58,3 +59,14 @@ fun ImageView.setDrawableOrHide(drawableRes: Drawable?) {
     setImageDrawable(drawableRes)
     isVisible = drawableRes != null
 }
+
+fun TextView.setText(text: String?, nullVisibility: Int, vararg relatedViews: View) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/extensions/ViewExtensions.kt:63]: I do not know why my comment is not referring to this line, nor why I cannot edit the comment. The nit is about this extension.

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt (hunk 6 above)
Comment (ouchadam) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/pushers/DefaultPushersService.kt]: makes sense, will do :+1:

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushers/PushersService.kt (hunk 2 above)
Comment (ouchadam) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/pushers/PushersService.kt]: after speaking offline, I've opted for adding a 3rd method for removing `Pushers` 4482cbdaa649ff299044a064fa1a9957a3db1819 this has the benefit of avoiding redundancy in checking the appId if we already have a `pusher` instance and we can have dedicated symmetrical http/email removal functions for convenience

File: vector/src/main/java/im/vector/app/core/pushers/PushersManager.kt [hunk 17]
```diff
@@ -75,12 +75,12 @@ class PushersManager @Inject constructor(
 
     suspend fun unregisterEmailPusher(email: String) {
         val currentSession = activeSessionHolder.getSafeActiveSession() ?: return
-        currentSession.removePusher(email, appId = "m.email")
+        currentSession.removeEmailPusher(email)
     }
 
     suspend fun unregisterPusher(pushKey: String) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/pushers/PushersManager.kt:81]: Nit: rename to `unregisterHttpPusher` for parity?

**Review by bmarty — APPROVED** [review]
Thanks for the update, LGTM!
